collector/src/main.rs

```diff
@@ -486,6 +486,85 @@ fn get_local_toolchain(
     Ok((rustc, rustdoc, cargo))
 }
 
+fn generate_cachegrind_diffs(
+    id1: &str,
+    id2: &str,
+    out_dir: &Path,
+    benchmarks: &[Benchmark],
+    build_kinds: &[BuildKind],
+    scenario_kinds: &[ScenarioKind],
+    errors: &mut BenchmarkErrors,
+) {
+    for benchmark in benchmarks {
+        for &build_kind in build_kinds {
+            for &scenario_kind in scenario_kinds {
+                if let ScenarioKind::IncrPatched = scenario_kind {
+                    continue;
+                }
+                let filename = |prefix, id| {
+                    format!(
+                        "{}-{}-{}-{:?}-{:?}",
+                        prefix, id, benchmark.name, build_kind, scenario_kind
+                    )
+                };
+                let id_diff = format!("{}-{}", id1, id2);
+                let cgout1 = out_dir.join(filename("cgout", id1));
+                let cgout2 = out_dir.join(filename("cgout", id2));
+                let cgdiff = out_dir.join(filename("cgdiff", &id_diff));
+                let cgann = out_dir.join(filename("cgann", &id_diff));
+
+                if let Err(e) = cg_diff(&cgout1, &cgout2, &cgdiff) {
+                    errors.incr();
+                    eprintln!("collector error: {:?}", e);
+                    continue;
+                }
+                if let Err(e) = cg_annotate(&cgdiff, &cgann) {
+                    errors.incr();
+                    eprintln!("collector error: {:?}", e);
+                    continue;
+                }
+            }
+        }
+    }
+}
+
+/// Compares two Cachegrind output files using cg_diff and writes result to path.
+fn cg_diff(cgout1: &Path, cgout2: &Path, path: &Path) -> anyhow::Result<()> {
+    let output = Command::new("cg_diff")
+        .arg("--mod-filename=s#/rustc/[^/]*/##")
+        .arg("--mod-funcname=s/[.]llvm[.].*//")
+        .arg(cgout1)
+        .arg(cgout2)
+        .stderr(Stdio::inherit())
+        .output()
+        .context("failed to run `cg_diff`")?;
+
+    if !output.status.success() {
+        anyhow::bail!("failed to generate cachegrind diff");
+    }
+
+    fs::write(path, output.stdout).context("failed to write `cg_diff` output")?;
+
+    Ok(())
+}
+
+/// Post process Cachegrind output file and writes resutl to path.
+fn cg_annotate(cgout: &Path, path: &Path) -> anyhow::Result<()> {
+    let output = Command::new("cg_annotate")
+        .arg(cgout)
+        .stderr(Stdio::inherit())
+        .output()
+        .context("failed to run `cg_annotate`")?;
+
+    if !output.status.success() {
+        anyhow::bail!("failed to annotate cachegrind output");
+    }
+
+    fs::write(path, output.stdout).context("failed to write `cg_annotate` output")?;
```

Comment:
I also typically rustfilt the input to cg_annotate -- seems like a nice thing to add? I'm not sure it's necessary without symbol-mangling-version = v0, but I think that valgrind doesn't yet support that (at least not on all platforms).

Reply:
I have been using valgrind-3.17.0 which does have demangling support builtin based on libiberty (I think?). The only limitation is lack of support for symbols renamed during LTO, those with dot suffix, i.e., `.llvm.*`.

Reply:
Ubuntu 20.04 (latest LTS) only has 3.15, so I think we should try to avoid relying on the built-in demangling, I suspect it doesn't work. (Admittedly, I've been using a locally compiled version from git, so who knows...).

Reply:
Anyway, we can do this later if it becomes a problem for folks.